### ISSUE TITLE
Multiple authentication mechanisms simultaneously

### ISSUE THREAD

**Issue by tofflos** [eclipse-ee4j/soteria#222]
I'm curious whether there has been any talk of supporting multiple authentication mechanisms simultaneously? The use case would be writing a JAX-RS endpoint that provides an API, some HTML/JS that provides an UI on top that API, and then bundling them together as a single application. I would then like to offer BASIC authentication for "machines" interacting directly with the API and FORM-based authentication for "humans" interacting indirectly via the UI.

**Comment by arjantijms** [comment]
Yes, there have been multiple discussion about this.

Not just about having multiple authentication mechanisms next to each other, but also having them stacked (so that for e.g. for JAX-RS multiple ones are tried automatically), as well as allowing users to select one (for things like giving users a chose to login with facebook and login with email).

Darran from JBoss even drafted a detailed proposal for the stacked ones.

Your question basically comes down to installing authentication mechanisms per URL pattern, which has been discussed in less details, but certainly has been mentioned.

This is in scope for EE Security 1.1, so I'm hoping we're able to get to it ;)

**Comment by tofflos** [comment]
Thank you for the response. I found the issue and mention in https://github.com/javaee/security-spec/issues/37:

> Note that this feature is distinct from having different authentication mechanisms for different resources, like e.g. FORM for the web based UI part at /app/* of an application and BASIC for the REST part at say /rest/*.

Do you think it would be a good idea to create a dedicated issue for authentication-mechanism-per-URL-pattern?

I'm also curious whether the JSR or the RI is actively looking for contributors? I'm not great at application server internals but I'm pretty good at communication and evaluation. That could translate into unit/integration testing, proofreading and writing of documentation.

**Comment by arjantijms** [comment]
>Do you think it would be a good idea to create a dedicated issue for authentication-mechanism-per-URL-pattern?

Yes, absolutely. Best place for the issue would be https://github.com/eclipse-ee4j/security-api

That's the API project, and authentication-mechanism-per-URL-pattern would be an API addition first.

> That could translate into unit/integration testing, proofreading and writing of documentation.

Sure, that's always needed. Basically just creating issues and providing comments (feedback) is already a contribution ;)

**Comment by ggam** [comment]
@tofflos please join the Soteria and Enterprise Security mailing lists if
you haven't dond already. Most discussions take place there. It will be
great to have you too!
